src/cbmc/cbmc_solvers.cpp

```diff
@@ -34,7 +34,8 @@ Author: Daniel Kroening, kroening@kroening.com
 /// \return An smt2_dect::solvert giving the solver to use.
 smt2_dect::solvert cbmc_solverst::get_smt2_solver_type() const
 {
-  assert(options.get_bool_option("smt2"));
+  // we shouldn't get here if this option isn't set
+  PRECONDITION(options.get_bool_option("smt2"));
```

Comment:
Hm this is not really a precondition of the function. I think in general it is fine to use `INVARIANT` even as the first statement in a function. But in the rest of the cbmc code base there are lots of `PRECONDITION` statements too that IMO should be `INVARIANTS`s instead.

Reply:
Can you explain what you mean by "not a precondition"? While the value in here isn't really needed for anything in the function, it's certainly something that should be true when we enter this function right?

Reply:
Hm I'm probably interpreting preconditions to narrowly. I basically took the view that a precondition is something that absolutely needs to hold when calling a function for it not to crash or produce garbage output. Like a function that unconditionally dereferences a pointer would have `ptr != nullptr` as a precondition, or a function that converts a string representing a binary number to an integer needs to get a string that contains only characters `'0'` and `'1'` (if it does not have built-in checks for this).

But currently in the codebase we have several preconditions that are more of the kind "here's something that happens to hold when we enter this function as it's currently used but it's not necessarily related to the function itself". Technically the functions could be used in different ways in the future and for those uses the preconditions might not apply anymore.

But the present case above seems somewhat like a corner case so I don't mind if we leave it as a `PRECONDITION`.

Reply:
Yeah, but we should probably dig up what the intended meaning is and get everyone to agree to it.